// @luckyorange/cli/src/lucky-docs.js
#!/usr/bin/env node

const program = require('commander')
const shell = require('shelljs')
const Listr = require('listr')
const execa = require('execa')
const chalk = require('chalk')

// Options
const DOCS_DIR = './docs'
const CONTRIBUTE_URL = 'https://docs.luckyorange.rocks/contribute.html'

program
  .version('1.0.0')
  .parse(process.argv)

// Clear console
process.stdout.write('\033c')

const tasks = new Listr([
  {
    title: 'Installing necessary dependencies',
    task: () => execa('npm i -D vuepress @luckyorange/vuepress-theme-docs', { shell: true })
  },
  {
    title: 'Creating docs directory',
    task: () => shell.mkdir('-p', `${DOCS_DIR}/.vuepress`)
  },
  {
    title: 'Adding docs directory to npmignore',
    task: () => {
      const npmignore = shell.cat('.npmignore')
      if (!npmignore.includes('docs')) {
        shell.ShellString('docs').toEnd('.npmignore')
      }
    }
  },
  {
    title: 'Creating Vuepress config',
    task: () => {
      const pkgJson = JSON.parse(shell.cat('package.json'))
      const name = pkgJson.name.replace('@luckyorange/', '')

      if (!shell.cat(`${DOCS_DIR}/README.md`).length) {
        shell.touch(`${DOCS_DIR}/README.md`)
      }

      if (!shell.cat(`${DOCS_DIR}/.vuepress/config.js`).length) {
shell.ShellString(
`module.exports = {
  title: '${name}',
  base: '/repos/${name}/',
  dest: 'docs/dist',
  theme: '@luckyorange/docs',
  themeConfig: {
    lastUpdated: 'Last updated',
    repo: 'luckyorange/${name}',
    docsDir: 'docs',
    editLinks: true,
    sidebar: 'auto'
  }
}
`).to(`${DOCS_DIR}/.vuepress/config.js`)
      }
    }
  },
  {
    title: 'Adding additional NPM scripts',
    task () {
      const pkgJson = JSON.parse(shell.cat('package.json'))

      pkgJson.scripts['build:docs'] = 'vuepress build docs'
      pkgJson.scripts['dev:docs'] = 'vuepress dev docs'

      shell.ShellString(JSON.stringify(pkgJson, null, 2)).to('package.json')
    }
  }
])

// Run the tasks
tasks.run().then(() => {
  console.log('\n')
  console.log(chalk.green(`  Docs website set up successfully in ${DOCS_DIR}!`))
  console.log(chalk.green(`  Get started by running: npm run dev:docs`))
  console.log(chalk.green(`  Or, learn more about building docs at ${CONTRIBUTE_URL}`))
  console.log('\n')
}).catch(() => {
  console.log('\n')
})
