// @luckyorange/cli/src/lucky-create-templates/integration/core/webpack.config.js
const fs = require('fs')
const path = require('path')

const joi = require('@hapi/joi')
const rimraf = require('rimraf')

const config = require('../integration.config')

const demoDir = 'demo/build'
const distDir = 'dist'

const devtool = process.env.NODE_ENV === 'production'
  ? false
  : 'inline-source-map'
const mode = process.env.NODE_ENV === 'production'
  ? 'production'
  : 'development'
const outputPath = process.env.DEMO === 'true'
  ? path.resolve(__dirname, demoDir)
  : path.resolve(__dirname, distDir)

function buildConfig () {
  const configObject = {
    id: config.id,
    settingsSchema: joi.describe(config.settings(joi))
  }
  fs.mkdirSync(
    `${path.resolve(__dirname, demoDir)}`
  )
  fs.writeFileSync(
    `${path.resolve(__dirname, demoDir)}/config.json`,
    JSON.stringify(configObject)
  )
}

rimraf.sync(outputPath)
if (process.env.DEMO) buildConfig()

module.exports = {
  mode,
  entry: `./src/${config.id}.js`,
  output: {
    filename: 'main.js',
    path: outputPath
  },
  devtool
}
