// @luckyorange/cli/src/lucky-create-templates/integration/core/demo/components/demo-controls.mjs
export default {
  model: {
    prop: 'integrationScriptLoaded',
    event: 'loadScript'
  },
  props: {
    controls: Array,
    integrationScriptLoaded: Boolean
  },
  data () {
    return {
      integrationScriptEl: null
    }
  },
  methods: {
    loadIntegrationScript () {
      const frameDocument = this.$refs.scriptFrame.contentDocument
      const scriptEl = frameDocument.createElement('script')
      scriptEl.src = '/build/main.js'
      frameDocument.head.appendChild(scriptEl)
      this.integrationScriptEl = scriptEl
      this.$emit('loadScript', true)
    }
  },
  template: `
    <section>
      <h1>Controls</h1>

      <section>
        <h1>Script Loader</h1>
        <button
          v-if="!integrationScriptLoaded"
          id="load-integration-script"
          @click="loadIntegrationScript"
          >Load integration script</button>
        <p v-else>Script loaded</p>
        <iframe
          ref="scriptFrame"
          src="about:blank"
          style="display: none;"
          ></iframe>
      </section>

      <section>
        <h1>Custom Controls</h1>
        <ul>
          <li
            v-for="control in controls"
            >
            <button
              :id="control.id"
              @click="control.handler"
              >
              \{{ control.name }}
            </button>
          </li>
        </ul>
      </section>
    </section>
  `
}
