// @luckyorange/cli/src/lucky-create-templates/integration/core/demo/modules/app.mjs
import Vue from 'https://cdnjs.cloudflare.com/ajax/libs/vue/2.6.10/vue.esm.browser.js'

import DemoControls from '/components/demo-controls.mjs'
import DemoLibraries from '/components/demo-libraries.mjs'
import DemoMocks from '/components/demo-mocks.mjs'
import DemoOutput from '/components/demo-output.mjs'
import DemoSettings from '/components/demo-settings.mjs'

export function createApp ({
  controls,
  libraries,
  mocks,
  name,
  settings,
  settingsSchema
}) {
  return new Vue({
    el: '#demo-app',
    components: {
      DemoControls,
      DemoLibraries,
      DemoMocks,
      DemoOutput,
      DemoSettings
    },
    data: {
      controls,
      integrationScriptLoaded: false,
      libraries,
      libraryOutput: null,
      mocks,
      mockOutput: null,
      name,
      output: [],
      settings,
      settingsSchema
    },
    watch: {
      libraryOutput (newOutput) {
        this.output.push(newOutput)
      },

      mockOutput (newOutput) {
        this.output.push(newOutput)
      },

      settings: {
        handler (newValue) {
          const settingsIndex = window._loIntegrationSettings.findIndex(
            setting => setting.name === this.name
          )
          window._loIntegrationSettings[settingsIndex] = newValue
        },
        deep: true
      }
    },
    template: `
      <section>
        <h1>Integration Demo</h1>
        <demo-settings
          v-model="settings"
          :settings-schema="settingsSchema"
          />
        <demo-libraries
          v-model="libraryOutput"
          :libraries="libraries"
          />
        <demo-mocks
          v-model="mockOutput"
          :mocks="mocks"
          />
        <demo-controls
          v-model="integrationScriptLoaded"
          :controls="controls"
          />
        <demo-output v-model="output" />
      </section>
    `
  })
}
